chapters/en/slides/module3/module3_02.Rmd

```diff
@@ -0,0 +1,308 @@
+    'runtime',
+    as_=['runtime', 'density'])  # Give a name to the KDE values, which we can use when plotting
+ .mark_area().encode(
+    x='runtime',
+    y='density:Q'))
+```
+
+Notes:
+To visualize a density plot,
+we need to do two things:
+
+1. Calculate the KDE (adding the individual kernels and summing them together)
+2. Plot a line or area mark for the newly calculated KDE values.
+
+In Altair,
+these operations are done in two explicit steps,
+using `transform_density` for the calculation.
+We could also have calculated this as a new column in our pandas data frame
+and then plotted that column,
+but it is convenient to do both the steps in Altair.
+
+The new variable we create is called `'density'`,
+and since this is not part of the pandas data frame,
+Altair cannot ask pandas which data type it is.
+Therefore,
+we need to add `':Q'` to indicate that the density has quantitative (numeric) values.
+
+---
+
+## Creating a grouped density plot requires an explicit density grouping
+
+```{python}
+(alt.Chart(movies).transform_density(
+     'runtime',
+     groupby=['genre'],
+     as_=['runtime', 'density'])
+ .mark_area().encode(
+     x='runtime',
+     y='density:Q',
+     color='genre'))
+```
+
+Notes:
+If we want to split and color the densities
+by a categorical variable,
+we need to add an explicit `groupby` column
+when calculating the density.
+This computes one density 
+per color variable (`genre` in this case),
+that will then be colored accordingly
+since we used the same variable for the color argument.
+
+In this plot,
+it is quite easy to see the relationship between movie lengths of different genres.
+The history genre has the longest running movies
+as well as the most variation
+whereas most animation moves are shorter
+and with less spread within the genre.
+
+There are still a few things we could tweak to improve this plot!
+
+---
+
+## Creating smoother density plots with more calculation steps
+
+```{python}
+(alt.Chart(movies).transform_density(
+     'runtime',
+     groupby=['genre'],
+     as_=['runtime', 'density'],
+     steps=200)
+ .mark_area().encode(
+     x='runtime',
+     y='density:Q',
+     color='genre'))
+```
+
+Notes:
+To make the curves smoother,
+we can increase the resolution
+over which the KDE is calculated
+by setting the step size to a higher number.
+
+Now there will be 200 points calculated,
+which gives the area a smoother appearance.
+
+---
+
+## Making the overlapping areas slightly transparent make them easier to interpret
+
+```{python}
+(alt.Chart(movies).transform_density(
+     'runtime',
+     groupby=['genre'],
+     as_=['runtime', 'density'],
+     steps=200)
+ .mark_area(opacity=0.4).encode(
+     x='runtime',
+     y='density:Q',
+     color='genre'))
+```
+
+Notes:
+The density areas don't suffer the same issues as the histograms when made transparent.
+The continuous solid shape is easier to follow
+even when it is semi transparent
+overlaps with the other areas.
```

Comment:
But eventually it might, if you had A LOT of densities - could you facet vertically to solve such a problem with altair?  Kind of like a ridge plot? If so, it would be nice to add a slide here showing that.

Reply:
You could but[ it is a bit hacky and rather involved](https://altair-viz.github.io/gallery/ridgeline_plot.html). I will see if I can add a simpler example.